### PR TITLE
restructured FStar.UInt

### PR DESCRIPTION
**Pull request by nickgian**
Hello all,

Changes are in FStar.UInt, and the new files FStar.UInt.Types, FStar.UInt.Base, FStar.UInt.Vectors.

1. FStar.UInt is divided into three files, FStar.UInt.Types which contains the definition of uint_t along with some auxiliary definitions, FStar.UInt.Base which contains the basic operations on uint_t such as addition, etc. and FStar.UInt.Vectors which include all the bitvector-related definitions such as logand, etc.
2. FStar.UInt.Types and FStar.UInt.Base use an interface file to hide some unnecessary details of the implementation. Probably the biggest gain from this comes from the fact that libraries such as FStar.Math.Lemmas are only used by the implementation of FStar.UInt.Types and not the interface. In vale we have noticed huge overheads when importing libraries such as Math.Lemmas (which we were inheriting from opening FStar.UInt). 
3. FStar.UInt.Vectors now only contains let definitions, not val definitions. I have not provided an interface for this file, if someone wants he can create one (it would be useful, but it was not my goal to do that, nor I am the right person to do it). In vale we use a special interface for this file that's Vale-specific, which allows us to import only what we need. This would not be possible if the val definitions were still in the file, as was the case in the original file.

I don't expect anything to break from these changes, FStar gets a green on CI, hopefully other projects sensitive to FStar.UInt still work.

### CODE REVIEW

**Comment by msprotz** [comment]
Can you clean up your pull request to exclude the ocaml-output directory? `git checkout master src/ocaml-output && git commit -m 'revert snapshot'`

**Comment by cpitclaudel** [comment]
Even better, you can rebase and remove these changes altogether

File: ulib/FStar.UInt.Base.fsti [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+let mul_mod (#n:nat) (a:uint_t n) (b:uint_t n) : Tot (uint_t n) =
+  (a * b) % (pow2 n)
+  
+val mul_div: #n:nat -> a:uint_t n -> b:uint_t n -> 
+  Tot (t:uint_t n{t = ((a * b) / (pow2 n))})
+
+(* Division primitives *)
+val div: #n:nat -> a:uint_t n -> b:uint_t n{b <> 0} -> Pure (uint_t n)
+  (requires (size (a / b) n))
+  (ensures (fun c -> b <> 0 ==> a / b = c))
```
Comment (tchajed) [inline ulib/FStar.UInt.Base.fsti]: Can you clean this up and remove `b <> 0` (it's already a refinement)?

File: ulib/FStar.UInt.Base.fsti [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+let gt #n (a:uint_t n) (b:uint_t n) : Tot bool = (a > b)
+let gte #n (a:uint_t n) (b:uint_t n) : Tot bool = (a >= b)
+let lt #n (a:uint_t n) (b:uint_t n) : Tot bool = (a < b)
+let lte #n (a:uint_t n) (b:uint_t n) : Tot bool = (a <= b)
+
+(* Casts *)
+let to_uint_t (m:nat) (a:int) : Tot (uint_t m) = a % pow2 m
+
+val incr_underspec: #n:nat -> a:uint_t n -> Pure (uint_t n)
+  (requires (b2t (a < max_int n)))
```
Comment (tchajed) [inline ulib/FStar.UInt.Base.fsti]: `b2t` is unnecessary

File: ulib/FStar.UInt.Base.fsti [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  (ensures (fun b -> a + 1 = b))
+
+val decr_underspec: #n:nat -> a:uint_t n -> Pure (uint_t n)
+  (requires (b2t (a > min_int n)))
```
Comment (tchajed) [inline ulib/FStar.UInt.Base.fsti]: same here

File: ulib/FStar.UInt.Base.fsti [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+val mul_underspec: #n:nat -> a:uint_t n -> b:uint_t n -> Pure (uint_t n)
+  (requires True)
+  (ensures (fun c ->
+    size (a * b) n ==> a * b = c))
+
+val div_underspec: #n:nat -> a:uint_t n -> b:uint_t n{b <> 0} -> Pure (uint_t n)
+  (requires True)
+  (ensures (fun c ->
+    (b <> 0 /\ size (a / b) n) ==> a / b = c))
```
Comment (tchajed) [inline ulib/FStar.UInt.Base.fsti]: Also don't need `b <> 0` in the postcondition here

File: ulib/FStar.UInt.Types.fsti [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+(* Constants *)
+let zero (n:nat) : Tot (uint_t n) = 0
+
+val pow2_n: #n:pos -> p:nat{p < n} -> Tot (r:uint_t n{r = pow2 p})
+
+#reset-options "--initial_fuel 1 --max_fuel 1"
+let one (n:pos) : Tot (uint_t n) = 1
+let ones (n:nat) : Tot (uint_t n) = max_int n
+//Just because let cannot be the last thing in the interface..
+val empty : Type0
```
Comment (tchajed) [inline ulib/FStar.UInt.Types.fsti]: This should be renamed `__empty` so it's less likely to conflict with user definitions.

File: ulib/FStar.UInt.Vectors.fst [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+	to_vec_lemma_2 #n num (from_vec (to_vec num))
+
+let from_vec_lemma_1 (#n:nat) (a:bv_t n) (b:bv_t n) :
+  Lemma (requires equal a b) (ensures from_vec a = from_vec b) = ()
+
+let from_vec_lemma_2 (#n:nat) (a:bv_t n) (b:bv_t n) :
+  Lemma (requires from_vec a = from_vec b) (ensures equal a b) = 
+  inverse_vec_lemma a; inverse_vec_lemma b
+
+open FStar.Math.Lemmas
```
Comment (tchajed) [inline ulib/FStar.UInt.Vectors.fst]: Can you move this to the top?

File: ulib/FStar.UInt.Types.fsti [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+   | 16 -> p=65536
+   | 31 -> p=2147483648
+   | 32 -> p=4294967296
+   | 63 -> p=9223372036854775808
+   | 64 -> p=18446744073709551616
+   | 128 -> p=0x100000000000000000000000000000000
+   | _  -> True))
+  [SMTPat (pow2 x)]
+  
+#reset-options "--initial_fuel 0 --max_fuel 0"
```
Comment (tchajed) [inline ulib/FStar.UInt.Types.fsti]: Don't need to set options in this `fsti` either.

File: ulib/FStar.UInt.Types.fsti [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+(* Machine integer type *)
+let uint_t (n:nat) = x:int{size x n}
+
+(* Constants *)
+let zero (n:nat) : Tot (uint_t n) = 0
+
+val pow2_n: #n:pos -> p:nat{p < n} -> Tot (r:uint_t n{r = pow2 p})
+
+#reset-options "--initial_fuel 1 --max_fuel 1"
```
Comment (tchajed) [inline ulib/FStar.UInt.Types.fsti]: Same here

File: ulib/FStar.UInt.Base.fsti [hunk 9]
```diff
@@ -0,0 +1,106 @@
+module FStar.UInt.Base
+open FStar.UInt.Types
+open FStar.Mul
+
+#reset-options "--initial_fuel 0 --max_fuel 0"
```
Comment (tchajed) [inline ulib/FStar.UInt.Base.fsti]: Can you remove this `#reset-options`? Setting options in an fsti has somewhat confusing semantics due to interleaving and we shouldn't need it in this file.

**Comment by nickgian** [comment]
For some reason UInt128 broke on the CI machine after those changes, it still works on my local machine with hints, and fails on a different lemma than the CI without hints... I assume it's the seemingly benign changes in the specifications of FStar.UInt.Base, will try reverting those.

**Comment by nickgian** [comment]
It was just hints that needed to be regenerated with z3rlimit_factor.

**Comment by nikswamy** [comment]
FYI, this pull request failed hacl-star regressions. I am investigating.

**Comment by nikswamy** [comment]
I've been trying in the `nik_uint` branch of hacl-star. 

It has 19 assertion failures. Here's the list if anyone is able to investigate

```$ grep "(Error)" dump.new
.\Hacl.Lib.LoadStore32.fst(45,50-45,52): (Error) assertion failed (see also D:\workspace\everest\FStar\ulib\FStar.UInt32.fst(65,12-65,32))
.\prf\Crypto.Symmetric.Cipher.fst(262,77-262,78): (Error) Subtyping check failed; expected type (n:FStar.UInt128.t{ Prims.b2t (FStar.UInt128.v n <
.\Spec.HMAC.fst(55,26-55,29): (Error) Subtyping check failed; expected type (b1:Spec.Lib.bytes{ Prims.b2t (FStar.Seq.Base.length ipad = FStar.Seq.Base.length b1) }); got type (key:Spec.Lib.bytes{ Prims.b2t (FStar.Seq.Base.length key = Spec.SHA2.size_block a) }) (see also .\Spec.HMAC.fst(19,39-19,60))
.\Spec.Chacha20_vec1.Lemmas.fst(1475,29-1475,55): (Error) Subtyping check failed; expected type Spec.Chacha20_vec.counter; got type Prims.int (see also D:\workspace\everest\FStar\ulib\FStar.UInt.Types.fsti(26,27-26,35))
.\Hacl.Spec.Bignum.Fsquare.fst(93,17-93,28): (Error) Subtyping check failed; expected type (b:Hacl.Bignum.Wide.t{ FStar.UInt.Types.size (Hacl.Bignum.Wide.v (d0 *^ r3) + Hacl.Bignum.Wide.v b)
.\Hacl.Spec.BignumQ.Mul.fst(573,2-573,8): (Error) assertion failed (see also .\Hacl.Spec.BignumQ.Mul.fst(573,8-573,26))
.\aead\Crypto.AEAD.Encrypt.Invariant.fst(81,45-81,50): (Error) Subtyping check failed; expected type (x:Crypto.Symmetric.PRF.domain i{ Prims.b2t (Mkdomain?.ctr x <^ Crypto.Symmetric.PRF.maxCtr i) }); got type Crypto.Symmetric.PRF.domain i (see also D:\workspace\everest\hacl-star\secure_api\prf\Crypto.Symmetric.PRF.fst(70,29-70,46))
.\aead\Crypto.AEAD.Enxor.Invariant.fst(99,31-99,42): (Error) Subtyping check failed; expected type Crypto.Symmetric.PRF.ctrT i; got type FStar.UInt32.t (see also D:\workspace\everest\hacl-star\secure_api\prf\Crypto.Symmetric.PRF.fst(65,21-65,35))
.\aead\Crypto.AEAD.Enxor.Invariant.fst(157,3-157,9): (Error) assertion failed (see also .\aead\Crypto.AEAD.Enxor.Invariant.fst(157,10-157,52))
.\Hacl.Bignum25519.fst(208,32-208,34): (Error) assertion failed (see also D:\workspace\everest\FStar\ulib\FStar.UInt64.fst(65,12-65,32))
.\aead\Crypto.AEAD.Wrappers.PRF.fst(58,37-58,49): (Error) Subtyping check failed; expected type Crypto.Symmetric.PRF.range (State?.mac_rgn t) i x; got type Crypto.Symmetric.UF1CMA.state (FStar.Pervasives.Native.Mktuple2 i (Mkdomain?.iv x)) (see also .\aead\Crypto.AEAD.Wrappers.PRF.fst(0,0-0,0))
.\Hacl.Impl.Ed25519.RecoverX.fsti(27,19-28,35): (Error) Subtyping check failed; expected type Prims.nat; got type Prims.int (see also D:\workspace\everest\FStar\bin\..\ulib\prims.fst(289,17-289,23))
.\Hacl.Impl.Ed25519.Sign.Steps.fst(246,46-246,80): (Error) could not prove post-condition
.\aead\Crypto.AEAD.Invariant.fst(475,43-475,56): (Error) assertion failed (see also D:\workspace\everest\FStar\ulib\FStar.UInt32.fst(45,12-45,32))
.\aead\Crypto.AEAD.Invariant.fst(780,21-780,77): (Error) Subtyping check failed; expected type Crypto.Symmetric.PRF.range rgn
.\Hacl.Hash.SHA2_512.fst(1175,93-1175,123): (Error) Subtyping check failed; expected type (j:Prims.nat{ Prims.b2t (FStar.UInt32.v Hacl.Hash.SHA2_512.pos_whash_w <= j &&
.\Hacl.Hash.SHA2_256.fst(976,55-976,57): (Error) assertion failed (see also D:\workspace\everest\FStar\ulib\FStar.UInt32.fst(65,12-65,32))
.\Hacl.Hash.SHA2_384.fst(884,2-884,72): (Error) Subtyping check failed; expected type (n:Hacl.Hash.SHA2_384.uint32_t{ Prims.b2t (FStar.UInt32.v n =
.\Hacl.Hash.SHA2_384.fst(1074,55-1074,57): (Error) assertion failed (see also D:\workspace\everest\FStar\ulib\FStar.UInt32.fst(65,12-65,32))```

**Comment by nikswamy** [comment]
Now down to 15 failures.

```
$ grep "(Error)" dump.new
.\prf\Crypto.Symmetric.Cipher.fst(179,0-283,5): (Error) Unknown assertion failed (SMT solver says: unknown because max. resource limit exceeded (fuel=0; ifuel=0; ))
.\Hacl.Lib.LoadStore32.fst(33,0-67,89): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=1; ))
.\Spec.HMAC.fst(47,0-70,16): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=0; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=1; ))
.\Hacl.Spec.BignumQ.Mul.fst(561,0-588,3): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=1; ))
.\Hacl.Hash.SHA2_512.fst(1169,0-1177,68): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=0; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=1; ))
.\aead\Crypto.AEAD.Encrypt.Invariant.fst(57,0-93,44): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=4; ifuel=2; ); SMT solver says: unknown because canceled (fuel=2; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=0; ))
.\aead\Crypto.AEAD.Invariant.fst(471,0-476,8): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=0; ifuel=0; ))
.\aead\Crypto.AEAD.Invariant.fst(764,0-781,41): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=1; ifuel=0; ))
.\aead\Crypto.AEAD.Wrappers.PRF.fst(50,0-63,19): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=4; ifuel=2; ); SMT solver says: unknown because canceled (fuel=2; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=0; ))
.\Hacl.Impl.Ed25519.RecoverX.fsti(18,0-34,8): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=1; ))
.\Hacl.Bignum25519.fst(200,0-217,55): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=1; ))
.\aead\Crypto.AEAD.Enxor.Invariant.fst(94,0-106,5): (Error) Unknown assertion failed (SMT solver says: unknown because canceled (fuel=0; ifuel=0; ))
.\aead\Crypto.AEAD.Enxor.Invariant.fst(157,3-157,9): (Error) assertion failed (see also .\aead\Crypto.AEAD.Enxor.Invariant.fst(157,10-157,52))
.\Hacl.Hash.SHA2_384.fst(883,0-884,72): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=1; ))
.\Hacl.Hash.SHA2_384.fst(1066,0-1159,48): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=1; ))
```

**Comment by nikswamy** [comment]
Getting there ... 

```.\prf\Crypto.Symmetric.Cipher.fst(235,6-237,65): (Error) Patterns are incomplete
.\aead\Crypto.AEAD.Enxor.Invariant.fst(157,3-157,9): (Error) assertion failed (see also .\aead\Crypto.AEAD.Enxor.Invariant.fst(157,10-157,52))
.\Hacl.Hash.SHA2_384.fst(1066,0-1159,48): (Error) Unknown assertion failed (SMT solver says: unknown because (resource limits reached) (fuel=0; ifuel=2; ); SMT solver says: unknown because canceled (fuel=0; ifuel=1; ))
```

**Comment by nikswamy** [comment]
Rather surprisingly, some of the brittleness here is because of the way kinds of unsat cores that Z3 seems to be reporting for some simple queries. For instance, in a large context, a query like this fails to replay:

```(assert (not (HasType (BoxInt 32) Prims.int)))```

It seems Z3 does not include `Prims.int_typing` in its unsat core because it somehow managed to find some other proof of this fact. But then excluding this axiom from the core causes the proof to fail forcing F* to ask Z3 to search for a proof again.

I'm trying out a patch on F* to unconditionally include axioms about primitive types when replaying hints. But, I wonder what it is about the new libraries that's causing this change in the proofs that Z3 is finding.

**Comment by ad-l** [comment]
@nikswamy What is the status of this branch?

**Comment by nikswamy** [comment]
This seems way out of date. I am closing it for now ... if anyone wants to revive it, please re-open.

Meanwhile, some of the ideas here are discussed in #1757

**Pull request closed without merging**